state/protocol/protocol_state/kvstore.go

```diff
@@ -105,25 +116,76 @@ type KVStoreMutator interface {
+// state machine.
+//
+// Orthogonal State Machines:
+// Orthogonality means that state machines can operate completely independently and work on disjoint
+// sub-states. They all consume the same inputs: the ordered sequence of Service Events sealed in one block.
+// In other words, each state machine S0, S1, ... has full visibility, but each draws their on independent
+// conclusions (maintain their own exclusive state).
+// This is a deliberate design choice. Thereby the default is favouring modularity and strong logical
+// independence. This is very beneficial for managing complexity in the long term.
+// We emphasize that this architecture choice does not prevent us of from implementing sequential state
+// machines for certain events. It is just that we would need to bundle these sequential state machines
+// into a composite state machine (conceptually a processing pipeline).
+//
+// Formally we write:
+//   - The overall protocol state 𝒫 is composed of disjoint substates 𝒫 = P0 ⊗ P1 ⊗  ... ⊗ Pj
+//   - For each state Pi, we have a dedicated state machine Si that exclusively operates on Pi
+//   - The state machines can be formalized as orthogonal regions of the composite state machine
+//     𝒮 = S0 ⊗ S1 ⊗  ... ⊗ Sj. (Technically, we represent the state machine by its state-transition
+//     function. All other details of the state machine are implicit.)
+//   - The state machine 𝒮 being in state 𝒫 and observing the input ξ = x0·x1·x2·..·xz will output
+//     state 𝒫'. To emphasize that a certain state machine 𝒮 exclusively operates on state 𝒫, we write
+//     𝒮[𝒫] = S0[P0] ⊗ S1[P1] ⊗ ... ⊗ Sj[Pj]
+//     Observing the events ξ the output state 𝒫' is
+//     𝒫' = 𝒮[𝒫](ξ) = S0[P0](ξ) ⊗ S1[P1](ξ) ⊗ ... ⊗ Sj[Pj](ξ) = P'0 ⊗ P'1 ⊗  ... ⊗ P'j
+//     Where each state machine Si individually generated the output state Si[Pi](ξ) = P'i
+//
+// The Protocol State is the framework, which orchestrates the orthogonal state machines,
+// feeds them with inputs, post-processes the outputs and overall manages state machines' life-cycle
+// from block to block. New key-value pairs and corresponding state machines can easily be added
+// by implementing the following interface (state machine) and adding a new entry to the KV store.
+type OrthogonalStoreStateMachine[P any] interface {
+
+	// Build returns:
+	//   - database updates necessary for persisting the updated protocol sub-state and its *dependencies*.
+	//     It may contain updates for the sub-state itself and for any dependency that is affected by the update.
+	//     Deferred updates must be applied in a transaction to ensure atomicity.
+	Build() protocol.DeferredDBUpdates
+
+	// ProcessUpdate processes an ordered list of sealed service events.
+	// Usually, each state machine performs filtering of relevant events and ignores all other events.
+	// No errors are expected during normal operations.
+	ProcessUpdate(orderedUpdates []flow.ServiceEvent) error
```

Comment:
this is an iteration on [Jordan's suggestion here](https://github.com/onflow/flow-go/pull/5616#discussion_r1549827439): 

```suggestion
	// EvolveState applies the state change(s) on sub-state P for the candidate block (under construction).
	// Information that potentially changes the Epoch state (compared to the parent block's state):
	//   - Service Events sealed in the candidate block
	//   - the candidate block's view (already provided at construction time)
	//
	// CAUTION: EvolveState MUST be called for all candidate blocks, even if `sealedServiceEvents` is empty!
	// This is because also the absence of expected service events by a certain view can also result in the
	// Epoch state changing (for example not having received the EpochCommit event for the next epoch, but
	// approaching the end of the current epoch).
	//
	// No errors are expected during normal operations.
	EvolveState(sealedServiceEvents []flow.ServiceEvent) error
```